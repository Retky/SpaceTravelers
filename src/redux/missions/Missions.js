import fetchMission from './api';

const GET_MISSION = 'MISSIONS_DISPLAYED';
const MISSION_JOINED = 'MISSION JOINED';
const MISSION_LEFT = 'MISSION LEFT';

const initialState = [];
let store = false;

export default function missionsReducer(state = initialState, action) {
  switch (action.type) {
    case GET_MISSION: return action.payload;
    case MISSION_JOINED: {
      const newState = state.map((mission) => {
        if (mission.mission_id !== action.payload.id) return mission;
        return { ...mission, reserved: action.payload.status };
      });
      return [...newState];
    }
    case MISSION_LEFT: {
      const newState = state.map((mission) => {
        if (mission.mission_id === action.payload.id) return mission;
        return { ...mission, reserved: action.payload.status };
      });
      return [...newState];
    }

    default: return state;
  }
}

// export const showMissions = () => async (dispatch) => {
//   let missions;
//   try {
//     if (store) return;
//     missions = await fetchMission();
//     dispatch({
//       type: GET_MISSION,
//       payload: missions,
//     });
//     store = true;
//   } catch (err) {
//     return err;
//   }
//   return missions;
// };

export const showMissions = () => async (dispatch) => {
  const missions = await fetchMission();
  if (store) return;
  dispatch({
    type: GET_MISSION,
    payload: missions,
  });
  store = true;
};

export const joinMissions = (id, status) => ({
  type: MISSION_JOINED,
  payload: { id, status: Boolean(Number(status)) },

});
