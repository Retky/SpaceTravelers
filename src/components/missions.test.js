import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/configureStore';
import Missions from './missions';

describe(' Missions Component Rendering ', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
