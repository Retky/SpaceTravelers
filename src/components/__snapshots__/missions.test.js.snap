// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Missions Component Rendering  renders correctly 1`] = `
<div>
  <hr />
  <table
    className="table"
  >
    <thead
      className="head"
    >
      <tr
        className="row"
      >
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
